src/librustc_lint/lib.rs

```diff
@@ -219,6 +219,11 @@ pub fn register_builtins(store: &mut lint::LintStore, sess: Option<&Session>) {
             reference: "issue #34537 <https://github.com/rust-lang/rust/issues/34537>",
             edition: None,
         },
+        FutureIncompatibleInfo {
+            id: LintId::of(EXTERNAL_PRIVATE_DEPENDENCY),
+            reference: "issue #44663 <https://github.com/rust-lang/rust/issues/44663>",
+            edition: None,
+        },
```

Comment:
Hmm, I don't think this should be marked as a future compatibility lint because it's hard to make it an error.
Errors assume precise rules, but the restrictions from the RFC are kind of heuristical and lint-y by nature.

Reply:
FWIW, `private_in_public` is also not going to be an error (see https://github.com/rust-lang/rust/issues/48054, it's going to be removed and replaced with more heuristic-based lints), it was just marked as future-compatibility historically.

Reply:
How will that interact with eventually preventing publishes to `crates.io`? Will `cargo publish` just set the lint to `deny`?

Reply:
>Also, the RFC states that this should explicitly become an error (at least when publishing to crates.io) at some point in the future.

The RFC also says

> In some situations, it can be necessary to allow private dependencies to become part of the public API. In that case one can permit this with #[allow(external_private_dependency)].

>Will cargo publish just set the lint to deny?

Yes, deny should work.

Reply:
Alternatively, the default could be flipped for `rustc` compared to Cargo, `--extern my_crate=PATH` would still mean public, and Cargo would generate `--extern my_crate:priv=PATH` for private dependencies instead.

Reply:
I agree that we probably don't want to place this under future-incompatible just yet. We'll want to iterate on the lint anyway before turning it on by default